llms/mlx_lm/evaluate.py

```diff
@@ -111,37 +96,26 @@ def _score_fn(self, inputs, tokenize=True, step_size=32):
         scores = mx.concatenate(scores, axis=1)
         is_greedy = mx.concatenate(is_greedy, axis=1)
 
-        return scores, mask.sum(axis=-1), is_greedy
-
-    def _loglikelihood(self, texts, score_spans=None, tokenize=True):
-        # sort by length to get batches with little padding.
-        sorted_indices = sorted(range(len(texts)), key=lambda i: -len(texts[i]))
-        sorted_inputs = [texts[sorted_indices[i]] for i in range(len(texts))]
-        sorted_spans = None
-        if score_spans is not None:
-            sorted_spans = [score_spans[sorted_indices[i]] for i in range(len(texts))]
+        return scores, lengths, is_greedy
 
+    def _loglikelihood(self, texts, score_spans=None):
         results = []
-        for i in tqdm(range(0, len(sorted_inputs), self._batch_size)):
-            batch = sorted_inputs[i : i + self._batch_size]
-            scores, length, is_greedy = self._score_fn(batch, tokenize=tokenize)
+        for i in tqdm(range(0, len(texts), self._batch_size)):
+            batch = texts[i : i + self._batch_size]
+            scores, length, is_greedy = self._score_fn(batch)
             for j in range(len(batch)):
```

Comment:
It's probably over optimization.. but having a bunch of evals in a loop like this is not such a good pattern. It hits latency really hard. Especially things like:

```
for i in range(len(lengths)):
  l = lengths[i].item()
```

Which does a kernel launch for the scatter + full GPU synch at each iteration. 

I know it probably will make little to no difference in runtime since this isn't the bottleneck, but perhaps still good to change it to set a good example.

Reply:
One way to do it would be:
- have a mask making function which takes `start_offset = mx.array, end_offset = mx.array` and makes the mask.
-  Then multiply `scores` and `is_greedy` by the mask. 
- Then sum along the time axis.
-  Then eval everything in one shot `mx.eval(scores, is_greedy)`
- Then convert them all to lists, zip and return

Reply:
Although it looks like keeping them as `mx.array` could be useful rather than converting back and forth here and at L205